### PR TITLE
fix(FP): Several false positive resolutions not supported by our automation

### LINKED ISSUE #4737: [FP]: spring-security-oauth2-authorization-server is mistaken for other Spring projects
### Package URl

pkg:maven/org.springframework.security/spring-security-oauth2-authorization-server@0.3.1

### CPE

`cpe:2.3:a:pivotal:spring_security_oauth:0.3.1:*:*:*:*:*:*:*`, `cpe:2.3:a:pivotal_software:spring_security:0.3.1:*:*:*:*:*:*:*`, `cpe:2.3:a:pivotal_software:spring_security_oauth:0.3.1:*:*:*:*:*:*:*`, `cpe:2.3:a:vmware:spring_security:0.3.1:*:*:*:*:*:*:*`

### CVE

CVE-2018-1260, CVE-2022-22978, CVE-2021-22112, CVE-2018-15758, CVE-2022-22969, CVE-2022-22976

### ODC Integration

{"label"=>"Gradle Plugin"}

### ODC Version

7.1.1

### Description

From the CVE descriptions this sounds to me as if CPEs for
- regular Spring Security are matched against the 0.3.1 version (it actually depends on 5.7.1: https://github.com/spring-projects/spring-authorization-server/blob/0.3.1/gradle.properties#L6)
- the old EOL legacy Security OAuth (https://github.com/spring-attic/spring-security-oauth) also match

### LINKED ISSUE #4820: [FP]: False Positive on Firebase GoogleUtilities
### Package URl

pkg:cocoapods/GoogleUtilities%2FAppDelegateSwizzler@7.7.0 

### CPE

cpe:2.3:a:app_project:app:7.7.0:*:*:*:*:*:*:*

### CVE

_No response_

### ODC Integration

{"label"=>"CLI"}

### ODC Version

7.1.2

### Problem We are using a Dependency checker to run vulnerabilities in our swift codebase using the below pod file: GoogleUtilities/AppDelegateSwizzler (7.7.0):

While running the dependency check, we found below errors

1. Cross-site scripting vulnerability in DeleGate 7.7.0 and 7.7.1 does not quote scripting commands within a "403 Forbidden" error page, which allows remote attackers to execute arbitrary Javascript on other clients via a URL that generates an error.

2. Multiple buffer overflows in DeleGate 7.7.0 through 7.8.1 allow remote attackers to execute arbitrary code, as demonstrated using a long USER command to the POP proxy.

3. Multiple implementations of the DNS protocol, including (1) Poslib 1.0.2-1 and earlier as used by Posadis, (2) Axis Network products before firmware 3.13, and (3) Men & Mice Suite 2.2x before 2.2.3 and 3.5.x before 3.5.2, allow remote attackers to cause a denial of service (CPU and network bandwidth consumption) by triggering a communications loop via (a) DNS query packets with localhost as a spoofed source address, or (b) a response packet that triggers a response packet.

4. Buffer overflow in the ssl_prcert function in the SSLway filter (sslway.c) for DeleGate 8.9.2 and earlier allows remote attackers to execute arbitrary code via a certificate with a long (1) subject or (2) issuer name field.

5. The DNS implementation in DeleGate 8.10.2 and earlier allows remote attackers to cause a denial of service via a compressed DNS packet with a label length byte with an incorrect offset, which could trigger an infinite loop.

6. Multiple buffer overflows in DeleGate before 8.11.1 may allow attackers to cause a denial of service or execute arbitrary code, possibly due to "overflows on arrays."

7. Multiple unspecified vulnerabilities in DeleGate 9.x before 9.0.6 and 8.x before 8.11.6 allow remote attackers to cause a denial of service via crafted DNS responses messages that cause (1) a buffer over-read or (2) infinite recursion, which can trigger a segmentation fault or invalid memory access, as demonstrated by the OUSPG PROTOS DNS test suite.

Same as: https://github.com/jeremylong/DependencyCheck/issues/2575

**Is this False Positive going to be fixed?**

### LINKED ISSUE #4827: [FP]: CVE-2021-22112 on spring-security-jwt
## Spring-security-jwt is incorrectly matching for CVEs related to spring-security.

*Package URl*
pkg:maven/org.springframework.security/spring-security-jwt@1.1.1.RELEASE

*CPE*
cpe:/a:pivotal_software:spring_security

*CVE*
https://nvd.nist.gov/vuln/detail/CVE-2021-22112

*ODC Version*
7.1.2

*Description*
Spring Security 5.4.x prior to 5.4.4, 5.3.x prior to 5.3.8.RELEASE, 5.2.x prior to 5.2.9.RELEASE, and older unsupported versions can fail to save the SecurityContext if it is changed more than once in a single request.A malicious user cannot cause the bug to happen (it must be programmed in). However, if the application's intent is to only allow the user to run with elevated privileges in a small portion of the application, the bug can be leveraged to extend those privileges to the rest of the application.

*False positive:*
Dependency spring-security-jwt:1.1.1.RELEASE.jar
Is detected as these Vulnerability IDs:
cpe:2.3:a:pivotal_software:spring_security:1.1.1:release:*:*:*:*:*:*

### LINKED ISSUE #4930: [FN]:  Apache parquet vulnerability not detected
### Package URl

pkg:maven/org.apache.parquet/parquet-format-structures@1.12.0

### CPE

cpe:2.3:a:apache:parquet-mr:*:*:*:*:*:*:*:*

### CVE

CVE-2021-41561

### ODC Integration

{"label"=>"Maven Plugin"}

### ODC Version

7.2.1

### Description

Apache parquet artifacts should be identified as apache:parquet-mr but I guess matching fails because parquet-mr is not used in artifact metadata (parquet-mr comes from the github repository name https://github.com/apache/parquet-mr to separate code from the format spec which is located at https://github.com/apache/parquet-format)

### LINKED ISSUE #4943: [FP]: Apache HBase Server compatibility matches Apache Http Server (httpd)
### Package URl

pkg:maven/org.apache.hbase/hbase-server@2.2.0

### CPE

cpe:2.3:a:apache:http_server:2.2.0:*:*:*:*:*:*:*

### CVE

_No response_

### ODC Integration

{"label"=>"Maven Plugin"}

### ODC Version

7.2.1

### Description

The Apache HBase server artifact is being mistaken for the Apache HTTP Server product (httpd)

### LINKED ISSUE #4945: [FP]: Apache Hive HBase handler matches Apache HBase
### Package URl

pkg:maven/org.apache.hive/hive-hbase-handler@3.1.1

### CPE

cpe:2.3:a:apache:hbase:3.1.1:*:*:*:*:*:*:*

### CVE

_No response_

### ODC Integration

{"label"=>"Maven Plugin"}

### ODC Version

7.2.1

### Description

Apache Hive handler for HBase is being mistaken for Apache HBase itself

### LINKED ISSUE #4951: [FP]: Apache Twill mistaken for Apache Hadoop
### Package URl

pkg:maven/org.apache.twill/twill-yarn@0.14.0

### CPE

cpe:2.3:a:apache:hadoop:0.14.0:*:*:*:*:*:*:*

### CVE

_No response_

### ODC Integration

{"label"=>"Maven Plugin"}

### ODC Version

7.2.1

### Description

The Apache Twill adapter for Hadoop Yarn is being mistaken for Apache Hadoop

### LINKED ISSUE #4953: [FP]: Apache Hive Spark Client matches Apache Spark
### Package URl

pkg:maven/org.apache.hive/hive-spark-client@3.1.1

### CPE

cpe:2.3:a:apache:spark:3.1.1:*:*:*:*:*:*:*

### CVE

_No response_

### ODC Integration

{"label"=>"Maven Plugin"}

### ODC Version

7.2.1

### Description

Apache Spark Client for Apache Hive is mistaken for Apache Spark itself (project is hosted by Apache Hive)

### LINKED ISSUE #4984: [FP]: Maven package zoomlayout detected as Zoom (meetings) software
### Package URl

pkg:maven/com.otaliastudios/zoomlayout@1.9.0?type=aar

### CPE

cpe:2.3:a:zoom:zoom:1.9.0:*:*:*:*:*:*:*

### CVE

CVE-2018-15715

### ODC Integration

{"label"=>"Gradle Plugin"}

### ODC Version

7.3.0

### Description

Hello,

The Maven package zoomlayout (https://mvnrepository.com/artifact/com.otaliastudios/zoomlayout) is being detected as:

- pkg:maven/com.otaliastudios/zoomlayout@1.9.0 (Confidence:Highest)
- **cpe:2.3:a:zoom:zoom:1.9.0:*:*:*:*:*:*:*  (Confidence:Low)**

Which leads to FP CVEs related to Windows/Mac OS/Linux versions of the Zoom software, e.g.
- [CVE-2018-15715](http://web.nvd.nist.gov/view/vuln/detail?vulnId=CVE-2018-15715) -> [cpe:2.3:a:zoom:zoom:*:*:*:*:*:mac_os_x:*:* versions up to (excluding) 4.1.34801.1116](https://web.nvd.nist.gov/view/vuln/search-results?adv_search=true&cves=on&cpe_version=cpe%3A%2Fa%3Azoom%3Azoom%3A%3A%3A%7E%7E%7Emac_os_x%7E%7E)
- [CVE-2017-15048](http://web.nvd.nist.gov/view/vuln/detail?vulnId=CVE-2017-15048) -> [cpe:2.3:a:zoom:zoom:*:*:*:*:*:linux:*:* versions up to (excluding) 2.0.115900.1201](https://web.nvd.nist.gov/view/vuln/search-results?adv_search=true&cves=on&cpe_version=cpe%3A%2Fa%3Azoom%3Azoom%3A%3A%3A%7E%7E%7Elinux%7E%7E)

Thank you.
Regards,

### LINKED ISSUE #5073: EcosystemCache gets used too early leading to potential False Negatives on freshly provisioned CVE DBs
**Describe the bug**
On a freshly provisioned database the CPEEntries are updated using the CPE Ecosystem Cache before the script that updates for known wrongly detected multi-ecosystem entries has been run

**Version of dependency-check used**
The problem occurs using version 7.3.3-SNAPSHOT

**To Reproduce**
1. Purge the database
2. Run update-only goal
3. Check the database and see that Apache Hadoop has entries with `native` ecosystem

**Expected behavior**
On a freshly provisioned database Apache Hadoop has `<null>` ecosystem to reflect that is is a multiple ecosystem product.

### PR DESCRIPTION
**Pull request by aikebah**
## Fixes Multiple FP Issues:
* fixes Issue #4999
* fixes Issue #4984
* fixes issue #4953
* fixes issue #4950 (including resolving the false-negatives for the proper ranger CPE (which was native in ecoSystemCache/cpeEntry)
* fixes issue #5073
* fixes issue #5045
* fixes issue #4827
* fixes issue #4945
* fixes issue #4943
* fixes issue #4820 
* fixes issue #4930
* fixes issue #4737
* fixes issue #4667

## Description of Change

Adds suppressions that fp-ops was not capable of generating and integrates the automatically generated suppressions up to now. Includes a code-fix for the ecosystem-cache (#5073)

## Have test cases been added to cover the new functionality?

no

### CODE REVIEW

File: core/src/main/resources/dependencycheck-base-suppression.xml [hunk 1]
```diff
@@ -5638,4 +5638,25 @@
...
+        <notes><![CDATA[
+        FP per issue #4984
+        ]]></notes>
+        <packageUrl regex="true">^pkg:maven/com\.otaliastudios/zoomlayout@.*$</packageUrl>
+        <cpe>cpe:/a:zoom:zoom</cpe>
+    </suppress>
+    <suppress base="true">
+        <notes><![CDATA[
+   FP per issue #4953 regex to match both hive-spark-client and the historical spark-client artifactIds
+   ]]></notes>
```
Comment (jeremylong) [inline core/src/main/resources/dependencycheck-base-suppression.xml]: Very minor suggestion on formatting the xml.
```suggestion
        FP per issue #4953 regex to match both hive-spark-client and the historical spark-client artifactIds
        ]]></notes>
```

**Comment by aikebah** [comment]
@jeremylong while fixing up the false negative for apache ranger I discovered a sequence-issue between the ecosystemcache updates and the database cleanup call, for which I'll include a commit shortly in this branch.
Currently validating that all tests still properly pass and then a final verification that now a freshly provisioned CVE database  has the proper ecosystem for Hadoop and Ranger.
The updates only take effect on the next actual update after the initial provisioning as the cleanDatabase executes (with the ecosystem updates) before the ecosystem cache has been updated to make these products MULTIPLE ecosystem.

**Comment by aikebah** [comment]
@jeremylong I'm done for now.... ready to integrate... including adding all generated suppressions up to this point (also marked in the generated suppressions file)

**Review by jeremylong — APPROVED** [review]

**Comment by jeremylong** [comment]
Thanks!

**Pull request merged**